From 3eef674a08ca904baff2e4d6a0f871e303d08e6e Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Mon, 26 Jan 2026 16:27:42 +0000
Subject: [PATCH] inspector: Enhance virt-v2v-inspector output with filesystems
 information

RHEL 7.0, 7.1 and (possibly*) 7.2 used XFS version 4.  New versions of
RHEL use XFS v5.

Support even for opening version 4 filesystems was removed in RHEL 10
(and will be removed altogether from the Linux kernel in 2030).  This
prevents virt-v2v conversions.

Therefore it's a good idea to be able to tell the XFS filesystem
version and print that in virt-inspector output, so that management
tools can warn about it.

Since we didn't have the virt-inspector <filesystems/> section at all
before, we have to add that whole section, sticking as close as
possible to the virt-inspector output.

Example output for a RHEL 7.0 guest:

  <?xml version='1.0' encoding='utf-8'?>
  <v2v-inspection>
    <!-- generated by virt-v2v-inspector 2.10.0local,libvirt -->
    <program>virt-v2v-inspector</program>
    ...
    <operatingsystem>
      <name>linux</name>
      ...
      <mountpoints>
        <mountpoint dev='/dev/sda3'>/</mountpoint>
        <mountpoint dev='/dev/sda1'>/boot</mountpoint>
      </mountpoints>
      <filesystems>
        <filesystem dev='/dev/sda1'>
          <type>ext4</type>
          <uuid>15e8838c-136e-4a1f-ac01-97b4fa6b0fe4</uuid>
        </filesystem>
        <filesystem dev='/dev/sda2'>
          <type>swap</type>
          <uuid>8e966377-86f1-45a1-bdff-28724530818d</uuid>
        </filesystem>
        <filesystem dev='/dev/sda3'>
          <type version='4'>xfs</type>
          <uuid>67ae8aeb-a9f4-4639-a201-c02b7dbb5d98</uuid>
        </filesystem>
      </filesystems>
    </operatingsystem>
  </v2v-inspection>

Example output for a RHEL 7.3 guest:

  <?xml version='1.0' encoding='utf-8'?>
  <v2v-inspection>
    <!-- generated by virt-v2v-inspector 2.10.0local,libvirt -->
    <program>virt-v2v-inspector</program>
    ...
    <operatingsystem>
      <name>linux</name>
      ...
      <mountpoints>
        <mountpoint dev='/dev/sda3'>/</mountpoint>
        <mountpoint dev='/dev/sda1'>/boot</mountpoint>
      </mountpoints>
      <filesystems>
        <filesystem dev='/dev/sda1'>
          <type>ext4</type>
          <uuid>e0467228-8727-482d-bd7a-6741885fe7ed</uuid>
        </filesystem>
        <filesystem dev='/dev/sda2'>
          <type>swap</type>
          <uuid>9c2e6658-c600-4fbd-8731-864ff987553c</uuid>
        </filesystem>
        <filesystem dev='/dev/sda3'>
          <type version='5'>xfs</type>
          <uuid>9692610c-0025-4a8c-b806-8a78de2ec2eb</uuid>
        </filesystem>
      </filesystems>
    </operatingsystem>
  </v2v-inspection>

To work this requires libguestfs >= 1.59.2 (with guestfs_xfs_info2).

The version field may be missing, for non-XFS filesystems, but also if
we cannot tell the XFS version for some reason.

* = The virt-builder rhel-7.2 image definitely uses XFS v4, but it may
have been built from an early (pre-)release of 7.2.  Later RHEL 7.2
seems to use XFS v5.

Fixes: https://issues.redhat.com/browse/RHEL-144075
(cherry picked from commit d6b2e643b174fff7d3442865da047bf3d3f9aac9)
---
 inspector/create_inspector_xml.ml | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/inspector/create_inspector_xml.ml b/inspector/create_inspector_xml.ml
index fb434227..a531cecc 100644
--- a/inspector/create_inspector_xml.ml
+++ b/inspector/create_inspector_xml.ml
@@ -126,6 +126,28 @@ let rec create_inspector_xml input_disks inspect target_meta =
   ) inspect.i_mountpoints;
   List.push_back os (e "mountpoints" [] !mps);
 
+  let fses = ref [] in
+  List.iter (
+    fun { fs_dev; fs_type; fs_version; fs_label; fs_uuid } ->
+      let fs = ref [] in
+      (match fs_type, fs_version with
+       | None, _ -> ()
+       | Some typ, None -> List.push_back fs (e "type" [] [PCData typ])
+       | Some typ, Some ver ->
+          List.push_back fs (e "type" [ "version", ver] [PCData typ])
+      );
+      (match fs_label with
+       | None -> ()
+       | Some label -> List.push_back fs (e "label" [] [PCData label])
+      );
+      (match fs_uuid with
+       | None -> ()
+       | Some uuid -> List.push_back fs (e "uuid" [] [PCData uuid])
+      );
+      List.push_back fses (e "filesystem" [ "dev", fs_dev] !fs)
+  ) inspect.i_filesystems;
+  List.push_back os (e "filesystems" [] !fses);
+
   List.push_back body (e "operatingsystem" [] !os);
 
   (* Construct the final document. *)
